crates/red_knot_python_semantic/src/lint.rs

```diff
@@ -0,0 +1,205 @@
+use itertools::Itertools;
+use ruff_db::diagnostic::{LintName, Severity};
+
+#[derive(Debug, Clone)]
+pub struct LintMetadata {
+    /// The unique identifier for the lint.
+    pub name: LintName,
+
+    /// A one-sentence summary of what the lint catches.
+    pub summary: &'static str,
+
+    /// An in depth explanation of the lint in markdown. Covers what the lint does, why it's bad and possible fixes.
+    ///
+    /// The documentation may require post-processing to be rendered correctly. For example, lines
+    /// might have leading or trailing whitespace that should be removed.
+    pub raw_documentation: &'static str,
+
+    /// The default level of the lint if the user doesn't specify one.
+    pub default_level: Level,
+
+    pub status: LintStatus,
+
+    /// Location where this lint is declared: `file_name:line`
+    pub source: &'static str,
+}
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Ord, PartialOrd)]
+pub enum Level {
+    /// The lint is disabled and should not run.
+    Ignore,
+
+    /// The lint is enabled and diagnostic should have a warning severity.
+    Warn,
+
+    /// The lint is enabled and diagnostics have an error severity.
+    Error,
+}
+
+impl Level {
+    pub const fn is_error(self) -> bool {
+        matches!(self, Level::Error)
+    }
+
+    pub const fn is_warn(self) -> bool {
+        matches!(self, Level::Warn)
+    }
+
+    pub const fn is_ignore(self) -> bool {
+        matches!(self, Level::Ignore)
+    }
+}
+
+impl TryFrom<Level> for Severity {
+    type Error = ();
+
+    fn try_from(level: Level) -> Result<Self, ()> {
+        match level {
+            Level::Ignore => Err(()),
+            Level::Warn => Ok(Severity::Warning),
+            Level::Error => Ok(Severity::Error),
+        }
+    }
+}
+
+impl LintMetadata {
+    pub fn name(&self) -> LintName {
+        self.name
+    }
+
+    pub fn summary(&self) -> &str {
+        self.summary
+    }
+
+    /// Returns the documentation line by line with leading and trailing whitespace removed.
```

Comment:
This says "leading and trailing whitespace removed", but that doesn't describe the behavior, it only removes a single leading space. I assume this is because the doc comments we get from the macro parser have a leading space on every line (the space between `///` and the text)?

Removing all leading whitespace from each line (as the wording currently implies) would be bad, as it would break many kinds of Markdown formatting.

```suggestion
    /// Returns the documentation line by line with one leading space and all trailing whitespace removed.
```

Reply:
> I assume this is because the doc comments we get from the macro parser have a leading space on every line (the space between /// and the text)?

Yes, this is the reason. We could remove the whitespace if we convert the macro to a proc-macro but I find proc-macros harder to maintain (and read), so I opted to do this at runtime.